File: Yandex/Practicum/Algorightms_and_Data_Structures/1_sprint/E.Самое_длинное_слово/solution.cpp
#include <iostream>
#include <string>

void solution(int L, std::string& sentence) {
    /*
    Time Complexity O(N)
    Memory Complexity O(L), where L is length of the longest word
    */
    std::string maxWord = "";
    int maxLength = 0;
    std::string curWord = "";
    int curLength = 0;
    for (int index = 0; index < L; ++index) {
        if (sentence[index] == ' ') {
            if (curLength > maxLength) {
                maxLength = curLength;
                maxWord = curWord;
            }
            curLength = 0;
            curWord = "";
        } else {
            ++curLength;
            curWord += sentence[index];
        }
    }
    if (curLength > maxLength) {
        maxLength = curLength;
        maxWord = curWord;
    }
    std::cout << maxWord << "\n" << maxLength << std::endl;
};


int main() {
    int L;
    std::cin >> L;
    std::cin.ignore();
    std::string sentence;
    std::getline(std::cin, sentence);
    solution(L, sentence);
}

File: Yandex/Practicum/Algorightms_and_Data_Structures/1_sprint/E.Самое_длинное_слово/solution.py
import sys

def solution(L: int, sentence: str) -> None:
    """
    Time Complexity O(N)
    Memory Complexity O(L), where L is length of the longest word
    """
    max_word: str = ''
    max_length: int = 0
    cur_word: list[str] = list()
    cur_length: int = 0
    for idx in range(L):
        if sentence[idx] == ' ':
            if cur_length > max_length:
                max_length = cur_length
                max_word = ''.join(cur_word)
            cur_length = 0
            cur_word.clear()
        else:
            cur_length += 1
            cur_word.append(sentence[idx])
    if cur_length > max_length:
        max_length = cur_length
        max_word = ''.join(cur_word)
    sys.stdout.write(max_word + '\n' + str(max_length) + '\n')

if __name__ == '__main__':
    L: int = int(sys.stdin.readline().rstrip())
    sentence: str = sys.stdin.readline().rstrip()
    solution(L, sentence)